compiler/src/dotty/tools/dotc/reporting/messages.scala

```diff
@@ -2987,12 +2987,26 @@ extends ReferenceMsg(CannotBeAccessedID):
     val where = if (ctx.owner.exists) i" from ${ctx.owner.enclosingClass}" else ""
     val whyNot = new StringBuffer
     for alt <- alts do
-      if alt.is(Protected) then
-        val cls = alt.owner.enclosingSubClass
-        val owner = if cls.exists then cls else alt.owner
-        val location = if owner.is(Final) then owner.showLocated else owner.showLocated + " or one of its subclasses"
-        whyNot.append(i"""
-            | Protected $alt can only be accessed from $location.""")
+      val cls = alt.owner.enclosingSubClass
+      val owner = if cls.exists then cls else alt.owner
+      val location: String =
+        if alt.is(Protected) then
+          def ownerOrInSubclass =
+            if owner.is(Final) then owner.showLocated
+            else owner.showLocated + " or one of its subclasses"
+          if !alt.privateWithin.exists then
+            ownerOrInSubclass
+          else if owner.is(Final) && owner.ownersIterator.contains(alt.privateWithin) then
+            alt.privateWithin.showLocated
+          else
+            alt.privateWithin.showLocated + ", or " + ownerOrInSubclass
```

Comment:
I think it makes more sense to only show private within when its different to the owner, e.g. otherwise we get this as seen in the test `protected[Foo] value Bar4 can only be accessed from class Foo, or class Foo or one of its subclasses.`

Reply:
Fixed and refactored a bit the code